rust/tw_internet_computer/src/protocol/identity.rs

```diff
@@ -0,0 +1,116 @@
+}
+
+/// Contains the signature and the associated DER-encoded public key from a call to
+/// [Identity::sign].
+pub struct IdentitySignature {
+    pub signature: Vec<u8>,
+    pub public_key: Vec<u8>,
+}
+
+/// An identity is a simple way to abstract away signing for envelopes.
+/// When creating a request to the IC, the sender and signature are required
+/// for authentication purposes. The sender is derived from a DER-encode public key
+/// and the signature is created using the private key.
+pub struct Identity {
+    private_key: PrivateKey,
+    der_encoded_public_key: Vec<u8>,
+}
+
+impl Identity {
+    /// Gets the public key and DER-encodes it and returns an Identity.
+    pub fn new(private_key: PrivateKey) -> Self {
+        let public_key = private_key.public();
+        let der_encoded_public_key = public_key.der_encoded();
+
+        Self {
+            private_key,
+            der_encoded_public_key,
+        }
+    }
+
+    /// Returns the principal of the private key.
+    /// Sender represents who is sending the request.
+    pub fn sender(&self) -> Principal {
+        Principal::self_authenticating(&self.der_encoded_public_key)
+    }
+
+    /// Signs the given content with the private key.
+    /// The signatures are encoded as the concatenation of the 32-byte big endian
+    /// encodings of the two values r and s.
+    ///
+    /// See: https://internetcomputer.org/docs/current/references/ic-interface-spec#ecdsa
+    pub fn sign(&self, content: H256) -> Result<IdentitySignature, SigningError> {
+        let signature = self
+            .private_key
+            .sign(content)
+            .map_err(SigningError::Failed)?;
+
+        let r = signature.r();
+        let s = signature.s();
+        let mut bytes = [0u8; 64];
+        bytes[..32].clone_from_slice(r.as_slice());
+        bytes[32..].clone_from_slice(s.as_slice());
+
+        let signature = IdentitySignature {
+            public_key: self.der_encoded_public_key.clone(),
+            signature: bytes.to_vec(), //Signature bytes
+        };
+
+        Ok(signature)
+    }
+}
+
+#[cfg(test)]
+mod test {
+
+    use tw_encoding::hex;
+
+    use super::*;
+
+    /// Test that the sender is derived from the private key.
+    #[test]
+    fn sender() {
+        let private_key = PrivateKey::try_from(
+            "227102911bb99ce7285a55f952800912b7d22ebeeeee59d77fc33a5d7c7080be",
+        )
+        .unwrap();
+        let identity = Identity::new(private_key);
+        let sender = identity.sender();
+        assert_eq!(
+            sender.to_text(),
+            "hpikg-6exdt-jn33w-ndty3-fc7jc-tl2lr-buih3-cs3y7-tftkp-sfp62-gqe"
+        )
+    }
+
+    /// Test signing with the identity.
+    #[test]
+    fn sign() {
+        let private_key = PrivateKey::try_from(
+            "227102911bb99ce7285a55f952800912b7d22ebeeeee59d77fc33a5d7c7080be",
+        )
+        .unwrap();
+        let der_encoded_public_key = private_key.public().der_encoded();
```

Comment:
Could you please consider implementing DER encoding for `secp256k1::PublicKey` manually? For example, we already have an implementation of DER encoding for `secp256k1::Signature`: [link](https://github.com/trustwallet/wallet-core/blob/dev/rust/tw_keypair/src/ecdsa/der.rs#L35).
It will allow us to avoid having `pem-rfc7468` in the dependency tree.

Reply:
Yes. I will take a look into this.

Reply:
I have pushed up a change for this. The `pem-rfc7468` crate has been removed.